`General`: Redesign the course overview

edkaya opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check all tasks by putting an x in the [ ] (don't: [x ], [ x], do: [x]). Remove not applicable tasks and do not leave them unchecked -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- Remove tasks that are not applicable for your PR. Please only put the PR into ready for review, if all relevant tasks are checked! -->
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable, and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Since the current design of the course overview page lacks user-friendliness, the pre-designed course cards will be im-
plemented to make the page appear cleaner and more structured. This implementation is a part of the new design for course overview page.

### Description
<!-- Describe your changes in detail -->
- Changed course card sizes and border design 
- Changed the way of displaying information of the score and next exercise
- Added Manage Course button for tutors, editors and instructors to navigate to the corresponding course management page

### Steps for Testing
<!-- Please describe in detail how reviewers can test your changes. Make sure to take all related features and views into account! Below is an example that you can refine. -->
Prerequisites:
- 1 Student
- 1 Tutor || Admin || Editor || Instructor

1. Log in to Artemis with student account
2. Verify that the course cards look good 
3. Verify that the course cards are responsive in case of size changes of page in width
4. Verify that you can navigate to the exercise by clicking on the exercise title under the "Next Exercise"
5. Log in to Artemis with Tutor || Admin || Editor || Instructor account
6. Verify that you can see manage course button 
7. Verify that manage course button works as expected

### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked
> Click on the badges to get to the test servers.

[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)](https://artemis-test1.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)](https://artemis-test2.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)](https://artemis-test3.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)](https://artemis-test4.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)](https://artemis-test5.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)](https://artemis-test6.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)](https://artemis-test9.artemis.cit.tum.de)

### Review Progress
<!-- Each PR should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. Remove the section if you did not change the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

Before 
<img width="1499" alt="Screenshot 2024-08-15 at 20 18 48" src="https://github.com/user-attachments/assets/04e37e64-7547-46c7-9a34-daeb0b9f1a46">

After 
<img width="1257" alt="Screenshot 2024-08-23 at 17 51 34" src="https://github.com/user-attachments/assets/adc9749e-d2ff-4b9e-8a62-09a75ff4cc40">

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
	- Enhanced course management interface with a new management button for direct access.
	- Introduced new methods to manage the state of the course overview.

- **User Interface Improvements**
	- Modernized the layout and design of the course card component with improved flexibility and visual clarity.
	- Enhanced pie chart representation and refined statistics display logic for better user feedback.
	- Improved responsive design across various screen sizes for better usability.
	- Updated HTML structure in the course overview component for better visual distinction and modernity.
	- Introduced a grid layout for course cards, enhancing adaptability and aesthetics.
	- Enhanced localization for both German and English versions of the student dashboard.

- **Bug Fixes**
	- Resolved issues related to state management of the course overview display.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

JohannesStoehr commented on src/main/webapp/app/overview/course-card.component.html:
```diff
@@ -2,101 +2,83 @@
+                            <div class="align-self-center">
+                                <div class="fw-bold text-secondary" jhiTranslate="artemisApp.studentDashboard.cardScore"></div>
+                                <div class="fw-bold">
+                                    <div>{{ totalAbsoluteScore }} / {{ totalReachableScore }} Points ({{ totalRelativeScore }}%)</div>
```
Please add a translation for this

JohannesStoehr commented on src/main/webapp/i18n/de/student-dashboard.json:
```diff
@@ -12,6 +12,9 @@
             "noStatistics": "Keine Statistik verfügbar",
             "cardNoExerciseLabel": "Keine Übung geplant",
             "cardExerciseLabel": "Nächste Übung:",
+            "cardNextUnit": "Nächste Einheit",
```
```suggestion
            "cardNextUnit": "Nächste Aufgabe",
```

JohannesStoehr commented on src/main/webapp/i18n/en/student-dashboard.json:
```diff
@@ -12,6 +12,9 @@
             "noStatistics": "No statistics available",
             "cardNoExerciseLabel": "No exercise planned",
             "cardExerciseLabel": "Next Exercise:",
+            "cardNextUnit": "Next Unit",
```
```suggestion
            "cardNextUnit": "Next Exercise",
```

b-fein commented:
Design comments looking at the screenshots:

- The headers of the cards now look like they take up more space than before. Would it be possible to make them a bit smaller again to reduce amount of blank space on the page?
- Artemis uses fairly rectangular shapes consistently throughout the UI. I don’t think the very rounded corners fit into this design. In the past there were even dedicated redesigns to make everything match the rectangular aesthetic (especially lectures and tutorials were not matching the other parts of the app if I remember correctly).

pzdr7 commented on src/main/webapp/app/overview/course-card.component.html:
```diff
@@ -2,101 +2,90 @@
+                    }
                 </div>
             </div>
+            @if (course.isAtLeastTutor || course.isAtLeastEditor || course.isAtLeastInstructor) {
```
Aren't instructors and editors already "at least tutors"?

```suggestion
            @if (course.isAtLeastTutor) {
```

pzdr7 commented on src/main/webapp/app/shared/layouts/main/main.component.scss:
```diff
@@ -0,0 +1,4 @@
+.course-overview {
+    background-color: var(--bs-body-bg);
+    padding-top: 0px;
```
```suggestion
    padding-top: 0;
```

or add the bootstrap class `pt-0` to the affected element

pzdr7 commented on src/main/webapp/app/overview/course-card.component.html:
```diff
@@ -2,101 +2,90 @@
                             </div>
                         }
                     </div>
+                    @if (exerciseCount > 0 && (totalReachableScore > 0 || totalAbsoluteScore > 0)) {
```
You might be able to simplify this by leaving out `> 0`, e.g. for exerciseCount. Can you check?

pzdr7 reviewed (changes requested):
Some small comments on the code

----

Regarding the design:

![image](https://github.com/user-attachments/assets/f982aa5d-2443-4afa-a899-9a9ca34307d6)

It seems that the image offsets the position of the course name. I'd prefer it if this wasn't the case. Maybe we can align the text differently and/or add a placeholder for courses without an image?

edkaya commented:
> Design comments looking at the screenshots:
> 
>     * The headers of the cards now look like they take up more space than before. Would it be possible to make them a bit smaller again to reduce amount of blank space on the page?
> 
>     * Having the vertical and horizontal spacing in the grid being different looks a bit weird to me. I think I’d prefer a consistent spacing like between rows at the moment also be used between items within the row.
> 
>     * Artemis uses fairly rectangular shapes consistently throughout the UI. I don’t think the very rounded corners fit into this design. In the past there were even dedicated redesigns to make everything match the rectangular aesthetic (especially lectures and tutorials were not matching the other parts of the app if I remember correctly).

Thanks a lot for your insights! We have adjusted design according to your remarks, which make sense in general. That's what we have changed: 
- The height of the header is reduced 
- We have more consistent spacing between the cards. Especially between the the header, cards, and the borders. Since we don't want to expand the cards too much like the current version (it happens if you have a larger screen and the grid tries to fit the content), we have a little bit more spacing between cards, which is a trade-off but expanding the card looks off and not good in general. 
- We reduced the rounded corners and adjusted to make it look consistent throughout Artemis.

janthoXO commented:
might be just a little detail, but isnt the green circle slightly slimmer than the red one
<img width="632" alt="image" src="https://github.com/user-attachments/assets/2386eee9-7cb9-4664-9a78-c266d2f2a2f9">

janthoXO reviewed (changes requested):
1. The "Next Exercise" does not communicate that it is clickable. As it has the same font and style as the points text and any "normal" text on the website

2. If the solution for 1. is to use the same style as "Manage Course" then the course card might look overloaded with text buttons. therefore I would change the "manage course" button to an icon button. probably something like an edit button with an pen icon.
In addition you should probably move this icon then somewhere to the top right of the card

<img width="515" alt="image" src="https://github.com/user-attachments/assets/62db467e-3b7f-4af7-bf53-5185f74479da">


Student View:
the padding on the bottom looks a bit too much. I know that's needed for the "manage course" button in instructor view but still looks a bit off. Using the icon pen button as said before, would solve this issue as well, as the padding could be reduced in both views
<img width="410" alt="image" src="https://github.com/user-attachments/assets/f573b82f-c979-42ec-a88a-1e19a3499a83">

anian03 reviewed (changes requested):
In general I do quite like the changes, great job! Just two things I noticed that we might want to change:

1. There appears to be some sort of double border (with non-matching corner radii) in dark mode, which looks strange
<img src="https://github.com/user-attachments/assets/d32c34d7-bc68-4a58-b177-76f3849e7909" width="200" alt="Double border">

2. The padding on the left and right side looks quite big if there is just one card (narrow screens like phone, Split View, …). Maybe can the cards just take up most of the width if just one is displayed per row?
<img src="https://github.com/user-attachments/assets/ae345948-57e5-47bd-8686-7d2a0406aae2" alt="Big horizontal empty space" width="300">

edkaya commented:
> 1. The "Next Exercise" does not communicate that it is clickable. As it has the same font and style as the points text and any "normal" text on the website
> 
>     2. If the solution for 1. is to use the same style as "Manage Course" then the course card might look overloaded with text buttons. therefore I would change the "manage course" button to an icon button. probably something like an edit button with an pen icon.
>        In addition you should probably move this icon then somewhere to the top right of the card
> 
> 
> <img alt="image" width="515" src="https://private-user-images.githubusercontent.com/95364200/361009204-62db467e-3b7f-4af7-bf53-5185f74479da.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ2ODAwNjQsIm5iZiI6MTcyNDY3OTc2NCwicGF0aCI6Ii85NTM2NDIwMC8zNjEwMDkyMDQtNjJkYjQ2N2UtM2I3Zi00YWY3LWJmNTMtNTE4NWY3NDQ3OWRhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI2VDEzNDI0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTljZmMxM2YzNzljNTUxZDAzNmUyZDUxMTY4ZjMyYzcyYWViYmVhYjI0NzI3MmY5MjgxNGYwMDM0NTNiYWRhMTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ZQRywpz8EQFI6L8GFYTLitAZ7k8Vu3uk2iZ75rh9ots">
> 
> Student View: the padding on the bottom looks a bit too much. I know that's needed for the "manage course" button in instructor view but still looks a bit off. Using the icon pen button as said before, would solve this issue as well, as the padding could be reduced in both views <img alt="image" width="410" src="https://private-user-images.githubusercontent.com/95364200/361009935-f573b82f-c979-42ec-a88a-1e19a3499a83.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ2ODAwNjQsIm5iZiI6MTcyNDY3OTc2NCwicGF0aCI6Ii85NTM2NDIwMC8zNjEwMDk5MzUtZjU3M2I4MmYtYzk3OS00MmVjLWE4OGEtMWUxOWEzNDk5YTgzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI2VDEzNDI0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk2NzE4MWU4YTBlYWYwMGY1MzkwNTM4MGZlN2JlODczOTg4MGVkYjM2NmM5MzhkYWFlYmFlNWFjZTVlMjFjMjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.qnE0_1pb58j3j6d87DsHxwCA-2wUkUlXZ8MfFQ4nKPk">

Thanks for your advices. You are right about the exercise title. I changed it to make it look clickable. Regarding your remaining points, we discussed these today with @rabeatwork and we came to conclusion that using an icon in the header will not be the best fit here, because it's not clear what will happen by clicking on this icon and there could be many possibilites from a user perspective. Users could understand that icon as course settings or editing the course title or something similar. Using a label like "Manage Course" will definetely convey the intended message here. Besides that, we may have some plans about the design of the header, and we can incorporate banner images. We'll need to consider the icon's visibility against those images, too.

edkaya commented:
> In general I do quite like the changes, great job! Just two things I noticed that we might want to change:
> 
>     1. There appears to be some sort of double border (with non-matching corner radii) in dark mode, which looks strange
> 
> 
> <img alt="Double border" width="200" src="https://private-user-images.githubusercontent.com/98647423/361044204-d32c34d7-bc68-4a58-b177-76f3849e7909.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ2ODA3OTgsIm5iZiI6MTcyNDY4MDQ5OCwicGF0aCI6Ii85ODY0NzQyMy8zNjEwNDQyMDQtZDMyYzM0ZDctYmM2OC00YTU4LWIxNzctNzZmMzg0OWU3OTA5LmpwZWc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgyNlQxMzU0NThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xMWYzNWJhMDAzZmVhNzg5MjA3OTQ4ZjY0ZjQwNDk2YmZjOGEwY2FhNjlhM2M4NjZhZTMwODc0MWI2MTNiNThmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.5-5IE3XS1wVxr-ZzsSxp7CWn-5ovoFhe00RvGKj86z4">
> 
>     2. The padding on the left and right side looks quite big if there is just one card (narrow screens like phone, Split View, …). Maybe can the cards just take up most of the width if just one is displayed per row?
> 
> 
> <img alt="Big horizontal empty space" width="300" src="https://private-user-images.githubusercontent.com/98647423/361044503-ae345948-57e5-47bd-8686-7d2a0406aae2.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ2ODA3OTgsIm5iZiI6MTcyNDY4MDQ5OCwicGF0aCI6Ii85ODY0NzQyMy8zNjEwNDQ1MDMtYWUzNDU5NDgtNTdlNS00N2JkLTg2ODYtN2QyYTA0MDZhYWUyLmpwZWc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgyNlQxMzU0NThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMDQzZGMyY2JjZDk1OTNiMmZiM2ZjMzBhMDhjNTA4YWE5ODg4ZjEwYjkwYTE5MWExOGY5YmI3NDJhOGQzNmE5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.DS6xh_Da7gv5y1_8NKU1XbfJEN0yJkt4kUomZXmPu6E">
> 
> Agree with the previous comment that it would be good to know that the next exercise is clickable – either via color and/or an arrow like "Manage Course".

Thanks, I fixed the border issue now. Regarding the spacing for one card, I think it will not lead to an issue on narrow screens, since this padding will not appear in these screens. You are right about the split view, but it seems an edge case to us, where the user views these cards in split view with only one card, because users generally use these split views after navigating to the course to see actual course content rather than course cards. Since we want to maintain consistency for the course card sizes across diferent screen sizes, we kind of make a trade-off here.

florian-glombik reviewed (changes requested):
The code looks quite clean, I only left some minor comments and suggestions, for my approve please address the comment in `src/main/webapp/app/overview/courses.component.html`

SimonEntholzer reviewed (approved):
The only thing that could be maybe improved is the overall course view layout when its very zoomed out, as the distances from the course cards increase a lot:
![image](https://github.com/user-attachments/assets/bf21f807-c51a-4a60-a2e9-c835748455ea)

But since people usually do not zoom out that far and don't even have that many courses, I think it's fine.

pzdr7 commented on src/main/webapp/app/course/manage/course-management.service.ts:
```diff
@@ -692,4 +695,12 @@ export class CourseManagementService {
         return this.http.get<number>(`${this.resourceUrl}/${courseId}/allowed-complaints?teamMode=${teamMode}`) ?? 0;
     }
+
+    setCourseOverviewBackground() {
+        this.courseOverviewSubject.next(true);
+    }
+
+    resetCourseOverviewBackground() {
+        this.courseOverviewSubject.next(false);
+    }
```
Maybe `enable` and `disable` would be a better terminology here than set and reset

pzdr7 commented on src/main/webapp/app/overview/course-card.component.html:
```diff
@@ -1,102 +1,100 @@
+                        <hr />
+                        <div class="fw-bold text-secondary" jhiTranslate="artemisApp.studentDashboard.cardExerciseLabel"></div>
+                        @if (nextRelevantExercise) {
+                            <a class="fw-bold btn-wrapper exercise-title exercise-guided-tour" [routerLink]="['/courses', course.id!, 'exercises', nextRelevantExercise.id!]">
+                                {{ nextRelevantExercise.title }}
+                            </a>
```
I think for `course.id` this is a problem of the template in general, rather than this specific location - basically the entire template relies on `course.id` not being `undefined` (everything but the outermost div). Personally I'd be fine with leaving the assertion for `course.id` as is. No strong opinion either way

pzdr7 reviewed (commented):
I can second what @SimonEntholzer said about the spacing of the cards. I have a very wide screen, which makes the cards look a bit small compared to the gaps/overall space. Probably an edge case, though

![image](https://github.com/user-attachments/assets/bed04944-3011-48c4-acde-2faf9b8f83a5)

anian03 reviewed (commented):
I think there is still something not matching with the inner and outer corner radius (see the black thing in the corner). That's not a big deal though, since it's not very noticeable. Will approve once the other comments are resolved :)
<img src="https://github.com/user-attachments/assets/a6ef80c3-bb0e-40b3-853f-01d1dd0c7333" width="200">

edkaya commented:
> sometimes the course overview loads extremly long. I'm not sure if its related to this PR but I never noticed it before so it might be related
> Screen.Recording.2024-08-28.at.10.19.41.mov

I don't have that issue when navigating to the course overview page. Are some of the other devs also experiencing this issue?

pzdr7 commented:
> > sometimes the course overview loads extremly long. I'm not sure if its related to this PR but I never noticed it before so it might be related
> > Screen.Recording.2024-08-28.at.10.19.41.mov
> 
> I don't have that issue when navigating to the course overview page. Are some of the other devs also experiencing this issue?

I couldn't reproduce this issue locally (tried develop & this branch). On TS3 (with admin access), there does appear to be some load time - but mostly on the course management side.

![image](https://github.com/user-attachments/assets/6d764fcc-5337-4c78-a5fb-e7c41b0f2508)

edkaya commented:
> I can second what @SimonEntholzer said about the spacing of the cards. I have a very wide screen, which makes the cards look a bit small compared to the gaps/overall space. Probably an edge case, though
> 
> ![image](https://private-user-images.githubusercontent.com/38403547/361968179-bed04944-3011-48c4-acde-2faf9b8f83a5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ4NzAwMDIsIm5iZiI6MTcyNDg2OTcwMiwicGF0aCI6Ii8zODQwMzU0Ny8zNjE5NjgxNzktYmVkMDQ5NDQtMzAxMS00OGM0LWFjZGUtMmZhZjliOGY4M2E1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI4VDE4MjgyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRjZmI0YjA2MGNkNmI5YWVjMTRiYjM0YTQ1NGVjYTZiN2I5OWRiMTI2NDY4MTA4ZWI1ZmQ2Nzg2MDFmODE4YmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PSXK7atuX9Mi7wOdL_Ermy6_oSfgnSxDtyyKpy208fs)

Thanks for your insights regarding the layout @pzdr7 and @SimonEntholzer. I added some breakpoints to adjust card sizes for larger screen to omit these edge cases as much as possible now 👍

pzdr7 reviewed (dismissed):
Approve code 👍

anian03 reviewed (dismissed):
Looks good to me :D

SimonEntholzer reviewed (commented):
One other thing I noticed: When going to the page, it first does not show the course pictures, and displays the title at the top.

https://github.com/user-attachments/assets/97e3bbc3-ef36-446f-a393-4e41e5ca8c3e

Would it be possible, to display the text at the correct place from the beginning?
And (or), like some websites do it, use some placeholder until the content has loaded, e.g. similar like this: https://experience.sap.com/fiori-design-web/placeholder-loading/

florian-glombik commented on src/main/webapp/app/overview/course-card.scss:
```diff
@@ -1,19 +1,31 @@
 .card {
-    min-height: 320px;
+    height: 283px;
+    min-width: 320px;
+    max-width: 400px;
     transition: transform 0.2s linear;
 
+    @media screen and (max-width: 768px) and (min-width: 478px) {
+        max-width: 550px !important;
+    }
+
+    @media screen and (min-width: 2000px) {
+        max-width: 500px !important;
+    }
+
+    @media screen and (min-width: 2700px) {
+        max-width: 600px !important;
+    }
```
Are these values based on trial and error? Do they have a deeper meaning?

Would it make sense to add an explanation here?

edkaya replied:
Not needed anymore, will remove them 👍

florian-glombik commented on src/main/webapp/app/overview/courses.component.html:
```diff
@@ -19,43 +19,51 @@ <h4 class="ms-2 fw-medium">{{ nextRelevantExam.title }}</h4>
+    }
+</div>
+
+<!-- course card template -->
```
I do not think that this comment adds a value, rather a comment for the future, does not need to be changed
```suggestion
```

florian-glombik reviewed (commented):
The code looks good to me except for `src/main/webapp/app/overview/course-card.scss` where it is not always clear to me on which basis we are choosing the values (e.g. `height: 283px` seems like a value that is inferred from an element, from which element?)

Could you have another look at this file and double check if it would make sense to add more comments on how we get to the px values / what they are meant to mean / supposed to achieve? 

If there are no useful comments that can be added I am happy to approve :)

edkaya commented:
> One other thing I noticed: When going to the page, it first does not show the course pictures, and displays the title at the top.
> 
>  Screencast.from.2024-08-29.10-15-40.mp4 
> Would it be possible, to display the text at the correct place from the beginning? And (or), like some websites do it, use some placeholder until the content has loaded, e.g. similar like this: https://experience.sap.com/fiori-design-web/placeholder-loading/

I think this happens due to high number of courses in test servers. I think it won't be a real case, since a student can have ideally 8-9 courses at most.

anian03 reviewed (approved):
Tested again on TS3, works as expected

pzdr7 reviewed (dismissed):
Reapprove after reviews

florian-glombik reviewed (dismissed):
Approve code, thanks for the changes!

janthoXO commented:
> I think there is still something not matching with the inner and outer corner radius (see the black thing in the corner). That's not a big deal though, since it's not very noticeable. Will approve once the other comments are resolved :) <img alt="" width="200" src="https://private-user-images.githubusercontent.com/98647423/361970362-a6ef80c3-bb0e-40b3-853f-01d1dd0c7333.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUyNjgxMzIsIm5iZiI6MTcyNTI2NzgzMiwicGF0aCI6Ii85ODY0NzQyMy8zNjE5NzAzNjItYTZlZjgwYzMtYmIwZS00MGIzLTg1M2YtMDFkMWRkMGM3MzMzLmpwZWc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwMlQwOTAzNTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iMzI2NGQwMDViNmExMGU0ZWNlNTI2MjljODkxOGM5Y2I3MTczMGMyMWQ0NWFhZGM5MzBiNzMzMDEzZTMwMzQxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.bBNjF_hC8geJf_wYKmzNCtBbYyyrAkLqVkGN8QrcGYM">

I would also say that this and the spacing on wide screens are the only issues left. After that you'll get an approve

edkaya commented:
> > I think there is still something not matching with the inner and outer corner radius (see the black thing in the corner). That's not a big deal though, since it's not very noticeable. Will approve once the other comments are resolved :) <img alt="" width="200" src="https://private-user-images.githubusercontent.com/98647423/361970362-a6ef80c3-bb0e-40b3-853f-01d1dd0c7333.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUyNjgxMzIsIm5iZiI6MTcyNTI2NzgzMiwicGF0aCI6Ii85ODY0NzQyMy8zNjE5NzAzNjItYTZlZjgwYzMtYmIwZS00MGIzLTg1M2YtMDFkMWRkMGM3MzMzLmpwZWc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkwMlQwOTAzNTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iMzI2NGQwMDViNmExMGU0ZWNlNTI2MjljODkxOGM5Y2I3MTczMGMyMWQ0NWFhZGM5MzBiNzMzMDEzZTMwMzQxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.bBNjF_hC8geJf_wYKmzNCtBbYyyrAkLqVkGN8QrcGYM">
> 
> I would also say that this and the spacing on wide screens are the only issues left. After that you'll get an approve

I fixed the border issue 👍 I normally expanded the cards for wide screens but I recently got some feedback from Ramona  that we don't want to expand the cards. Instead, we now display more cards in a row if the screen is larger. But I don't think we can have a super optimal solution for very wide screens (+3500 pixels) to decrease that spacing, since it is the intended behaviour now rather than an issue.

anian03 reviewed (approved):
Thanks for the changes! Tested & looks good

janthoXO reviewed (changes requested):
The border radius looks good now

The spacing is not satisfying yet:
<img width="1047" alt="image" src="https://github.com/user-attachments/assets/86fdf882-5579-46ff-a7e8-bc196cd714c9">

Another problem I just noticed, is that for some images the icon is not a circle but rather an elipsis
<img width="1012" alt="image" src="https://github.com/user-attachments/assets/0c88fd14-901b-4654-bfce-dea155238cfa">

edkaya commented:
> The border radius looks good now
> 
> The spacing is not satisfying yet: <img alt="image" width="1047" src="https://private-user-images.githubusercontent.com/95364200/363696440-86fdf882-5579-46ff-a7e8-bc196cd714c9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUyODg4MzAsIm5iZiI6MTcyNTI4ODUzMCwicGF0aCI6Ii85NTM2NDIwMC8zNjM2OTY0NDAtODZmZGY4ODItNTU3OS00NmZmLWE3ZTgtYmMxOTZjZDcxNGM5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTAyVDE0NDg1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA2YTU3NTkyYTMwNDIyYzg1NzZjYTNlODUwNzM3MWQ4MmNhYTgyZjljNjQ4N2I0ODUzMzAxOWFjNWMxM2U0OTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.KOqI74-B1fqKvlRKkvDLJvP3tu4kRoqjrYnneV2sZ6U">
> 
> Another problem I just noticed, is that for some images the icon is not a circle but rather an elipsis <img alt="image" width="1012" src="https://private-user-images.githubusercontent.com/95364200/363696170-0c88fd14-901b-4654-bfce-dea155238cfa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUyODg4MzAsIm5iZiI6MTcyNTI4ODUzMCwicGF0aCI6Ii85NTM2NDIwMC8zNjM2OTYxNzAtMGM4OGZkMTQtOTAxYi00NjU0LWJmY2UtZGVhMTU1MjM4Y2ZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTAyVDE0NDg1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY1NmI4MDZhZGY3Nzc3NjZmYTA3MDI0NjI5ZWVjMjBlMTViMGI5MDJmNjA1MzlkYTcwNjI4ZjU3YTk1NzIwNDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.B6KtKUIxaWKFGnca6plpRhNaDPpOzEmWxNBMzlkAaT8">

Thanks for your response. Regarding the spacing again, I now added some additional breakpoints for very large screens. We now display 9 cards at most for ~5000 pixels width, so the spacing should be decreased and fine now. I don't think we should take care of the case where the display is more than 5000 pixels, because the cards will be barely visible. 
<img width="1786" alt="Screenshot 2024-09-02 at 17 35 28" src="https://github.com/user-attachments/assets/1359a926-4b77-45cd-9a97-9b1535388705">

Regarding the icon issue, this is actually not related to this PR and I think something is wrong with the icon itself. Because it is also ellipsis in the current course card version: 
<img width="600" alt="Screenshot 2024-09-02 at 17 02 59" src="https://github.com/user-attachments/assets/be6f0478-208b-4c0c-b186-ee18d90d1969">

anian03 reviewed (dismissed):
Re-approve, still lgtm on TS3

SimonEntholzer reviewed (dismissed):
I think it looks good now :+1: tested on TS3

janthoXO commented:
> Thanks for your response. Regarding the spacing again, I now added some additional breakpoints for very large screens. We now display 9 cards at most for ~5000 pixels width, so the spacing should be decreased and fine now. I don't think we should take care of the case where the display is more than 5000 pixels, because the cards will be barely visible. 

### Spacing
I dont quite understand yet why we "hardcode" the break points. Is there no way in angular to just fit as many as possible in one row. Im only used to Flutter where there is a [Wrap Component](https://api.flutter.dev/flutter/widgets/Wrap-class.html) which does exactly that.
Another solution for the problem would be to hard code the maximum pixel width for the box containing all course card. so that with increasing screen width only the padding on the left and right side of the screen increases. Github does exactly that for example:
![image](https://github.com/user-attachments/assets/172c7c77-0f27-4d4a-bf43-7b16b4bf24e9)


> Regarding the icon issue, this is actually not related to this PR and I think something is wrong with the icon itself. Because it is also ellipsis in the current course card version

even if its not directly related to the PR, as this PR already touches the course cards, we could fix the issue. 
if you really don't want to fix it in this PR could you please create an issue and fix it in a separate PR.

florian-glombik reviewed (dismissed):
Approve code

I do not have a strong opinion on the suggestion https://github.com/ls1intum/Artemis/pull/9221#issuecomment-2325216957. If we can set a fixed max-width and a fixed padding I agree that this might be a better solution. If we need to adjust the padding at certain breakpoints I am not sure if it makes a difference.

edkaya reviewed (commented):
I used a different approach to omit using breakpoints. Instead of using columns and modifying them according to the breakpoints, I implemented with grid layout, which made the code more intuitive now.

anian03 reviewed (changes requested):
Hey, I just checked out the current version on my phone and noticed an issue with the layout: The course cards are wider than their container. Is it possible to fix this?
I think for these small screen sizes we could simply make the container take up the whole width, there is no advantage in having extra padding around it when there already is padding inside the container.
<img src="https://github.com/user-attachments/assets/f8d14287-e4f1-46a7-94e3-91263cefde76" width="250">

anian03 commented:
Thanks for the fix! Layout looks good on mobile now 👍

SimonEntholzer reviewed (approved):
re-approve, looks good on ts3

janthoXO reviewed (approved):
looks really good now. thanks for implementing the changes

florian-glombik commented on src/main/webapp/app/overview/course-card.scss:
```diff
@@ -1,5 +1,8 @@
 .card {
-    min-height: 320px;
+    // best possible height to fit header, information, and manage course button: 80px header + 200px card content
+    height: 283px;
```
80px + 200px = 280px

What are the 3 px related to?

edkaya replied:
That additional pixel is for the course manage button, I didnt write it explicitly because it was not the case for all cards 😅

edkaya replied:
But I can change this comment too, as I will modify some css properties of the course cards in my follow up PR

florian-glombik reviewed (commented):
Thanks for the comments explaining what happens in the CSS, there is only one comment that I do not quite understand 😅

florian-glombik reviewed (approved):
Approve code

pzdr7 reviewed (dismissed):
Reapprove code

anian03 reviewed (approved):
Tested on TS3, Score percentage has been moved

SimonEntholzer reviewed (approved):
Looks good on Ts3, reapprove :+1:

florian-glombik reviewed (approved):
Approve code

pzdr7 reviewed (approved):
Code

janthoXO reviewed (approved):
tested on ts1
